flow/layers/shader_mask_layer.cc

```diff
@@ -33,12 +33,23 @@ void ShaderMaskLayer::Preroll(PrerollContext* context, const SkMatrix& matrix) {
   Layer::AutoPrerollSaveLayerState save =
       Layer::AutoPrerollSaveLayerState::Create(context);
   ContainerLayer::Preroll(context, matrix);
+
+  if (render_count_ >= kMinimumRendersBeforeCachingFilterLayer) {
```

Comment:
> Good advice. ShaderMask just need to draw shader on the leaf_nodes_canvas, not do work on all its child layers. If caching child, It seems that just use raster cache to replace savelayer, I'm not sure whether it's worth to do, but will test it later.

I think it's worthwhile. You can think of the save layer as a special raster cache which cannot be reused by the next frame.

Reply:
It looks like this? I'm not sure `DrawMask` works fine  in all cases when `raster_cache->Draw(GetCacheableChild()...` is true if there is no savelayer.
```C++
void ShaderMaskLayer::Preroll(PrerollContext* context, const SkMatrix& matrix) {
  Layer::AutoPrerollSaveLayerState save =
      Layer::AutoPrerollSaveLayerState::Create(context);
  ContainerLayer::Preroll(context, matrix);

  if (render_count_ >= kMinimumRendersBeforeCachingFilterLayer) {
    TryToPrepareRasterCache(context, this, matrix);
  } else {
    render_count_++;
    TryToPrepareRasterCache(context, GetCacheableChild(), matrix);
  }
}

void ShaderMaskLayer::Paint(PaintContext& context) const {
  TRACE_EVENT0("flutter", "ShaderMaskLayer::Paint");
  FML_DCHECK(needs_painting(context));

  if (context.raster_cache) {
    if (context.raster_cache->Draw(this, *context.leaf_nodes_canvas)) {
      return;
    }

    AutoCachePaint cache_paint(context);
    if (context.raster_cache->Draw(GetCacheableChild(),
                                   *context.leaf_nodes_canvas,
                                   cache_paint.paint())) {
      DrawMask(context);
    }
  }

  AutoCachePaint cache_paint(context);
  Layer::AutoSaveLayer save = Layer::AutoSaveLayer::Create(
      context, paint_bounds(), cache_paint.paint());
  PaintChildren(context);

  DrawMask(context);
}

void ShaderMaskLayer::DrawMask(PaintContext& context) const {
  SkPaint paint;
  paint.setBlendMode(blend_mode_);
  paint.setShader(shader_);
  context.leaf_nodes_canvas->translate(mask_rect_.left(), mask_rect_.top());
  context.leaf_nodes_canvas->drawRect(
      SkRect::MakeWH(mask_rect_.width(), mask_rect_.height()), paint);
}
```

Reply:
> It looks like this? It will use child raster cache twice, then itself. But, when `context.raster_cache->Draw(GetCacheableChild()...` is true, I'm still not sure `DrawMask` will work fine in all cases if there is no savelayer created, because the last two instructions(translate/drawRect) seem work on savelayer according to original code.
> 
> ```c++
> void ShaderMaskLayer::Preroll(PrerollContext* context, const SkMatrix& matrix) {
>   Layer::AutoPrerollSaveLayerState save =
>       Layer::AutoPrerollSaveLayerState::Create(context);
>   ContainerLayer::Preroll(context, matrix);
> 
>   if (render_count_ >= kMinimumRendersBeforeCachingFilterLayer) {
>     TryToPrepareRasterCache(context, this, matrix);
>   } else {
>     render_count_++;
>     TryToPrepareRasterCache(context, GetCacheableChild(), matrix);
>   }
> }
> 
> void ShaderMaskLayer::Paint(PaintContext& context) const {
>   TRACE_EVENT0("flutter", "ShaderMaskLayer::Paint");
>   FML_DCHECK(needs_painting(context));
> 
>   if (context.raster_cache) {
>     if (context.raster_cache->Draw(this, *context.leaf_nodes_canvas)) {
>       return;
>     }
> 
>     AutoCachePaint cache_paint(context);
>     if (context.raster_cache->Draw(GetCacheableChild(),
>                                    *context.leaf_nodes_canvas,
>                                    cache_paint.paint())) {
>       DrawMask(context);
>       return;
>     }
>   }
> 
>   AutoCachePaint cache_paint(context);
>   Layer::AutoSaveLayer save = Layer::AutoSaveLayer::Create(
>       context, paint_bounds(), cache_paint.paint());
>   PaintChildren(context);
> 
>   DrawMask(context);
> }
> 
> void ShaderMaskLayer::DrawMask(PaintContext& context) const {
>   SkPaint paint;
>   paint.setBlendMode(blend_mode_);
>   paint.setShader(shader_);
>   context.leaf_nodes_canvas->translate(mask_rect_.left(), mask_rect_.top());
>   context.leaf_nodes_canvas->drawRect(
>       SkRect::MakeWH(mask_rect_.width(), mask_rect_.height()), paint);
> }
> ```

Also cc @flar, Could you have any more suggestions?  Thanks a lot.

Reply:
```
class ShadeMaskLayer : public MergedContainerLayer {....}

void Preroll(...) {
  if (render_count_ >= kMinimumRendersBeforeCachingFilterLayer) { 
    TryToPrepareRasterCache(context, this, matrix); } 
   else {  
    
     render_count_++; 
     
     transform_shade_masker_ = shader_->makeWithLocalMatrix(matrix);
     if (transform_shade_masker_) {
       TryToPrepareRasterCache(context, GetCacheableChild(), matrix);
     }
   } 
}


```

Reply:
The need for transform_filter in ImageFilterLayer is that the filter will be applied during the draw of the cached child - and the draw of the cached child is done in device space. So, we need to modify the filter so that it draws in device space what it would have drawn in the layer-local coordinate space.

The shader here is applied after the children are drawn and in the layer-local coordinate space. However, there is a problem in that it should be applied inside a saveLayer that drew the children so that it does not affect anything that lies under the children. So, the Paint method is wrong here.